lib/Mount/FolderStorageManager.php

```diff
@@ -106,11 +223,55 @@ public function getBaseStorageForFolder(int $folderId, ?FolderDefinition $folder
 
 	public function deleteStoragesForFolder(FolderDefinition $folder): void {
 		foreach (['files', 'trash', 'versions'] as $type) {
-			$storage = $this->getBaseStorageForFolder($folder->id, $folder, null, false, $type);
+			$storage = $this->getBaseStorageForFolder($folder->id, $folder->useSeparateStorage(), $folder, null, false, $type);
 			/** @var Cache $cache */
 			$cache = $storage->getCache();
 			$cache->clear();
 			$storage->rmdir('');
 		}
 	}
+
+	private function getObjectStorageKey(int $folderId): string {
+		$configs = $this->primaryObjectStoreConfig->getObjectStoreConfigs();
+		if ($this->primaryObjectStoreConfig->hasMultipleObjectStorages()) {
+			$configKey = 'object_store_key_' . $folderId;
+			$storageConfigKey = $this->appConfig->getValueString('groupfolders', $configKey);
+			if (!$storageConfigKey) {
+				$storageConfigKey = isset($configs['groupfolders']) ? $this->primaryObjectStoreConfig->resolveAlias('groupfolders') : $this->primaryObjectStoreConfig->resolveAlias('default');
+				$this->appConfig->setValueString('groupfolders', $configKey, $storageConfigKey);
+			}
+			return $storageConfigKey;
+		} else {
+			return 'default';
+		}
+	}
+
+	private function getObjectStorageBucket(int $folderId, array $objectStoreConfig): string {
+		$bucketKey = 'object_store_bucket_' . $folderId;
+		$bucket = $this->appConfig->getValueString('groupfolders', $bucketKey);
+		if (!$bucket) {
+			if (!isset($objectStoreConfig['arguments']['bucket'])) {
+				$objectStoreConfig['arguments']['bucket'] = '';
+			}
+			$bucket = $objectStoreConfig['arguments']['bucket'] . $this->calculateBucketNum((string)$folderId, $objectStoreConfig);
+
+			$this->appConfig->setValueString('groupfolders', $bucketKey, $bucket);
+		}
+		return $bucket;
+	}
+
+	// logic taken from OC\Files\ObjectStore\Mapper which we can't use because it requires an IUser
+	private function calculateBucketNum(string $key, array $objectStoreConfig): string {
+		$numBuckets = $objectStoreConfig['arguments']['num_buckets'] ?? 64;
+
+		// Get the bucket config and shift if provided.
+		// Allow us to prevent writing in old filled buckets
+		$minBucket = isset($objectStoreConfig['arguments']['min_bucket'])
+			? (int)$objectStoreConfig['arguments']['min_bucket']
+			: 0;
```

Comment:
```suggestion
		$minBucket = $objectStoreConfig['arguments']['min_bucket'] ?: 0;
```
I think this should also work :thinking: